### PR TITLE
NIFI-16028 - Support nested field paths in PutElasticsearchJson

### PR DESCRIPTION
**Pull request by agturley**
# Summary

[NIFI-16028](https://issues.apache.org/jira/browse/NIFI-16028)

Allow the Index Field, Identifier Field, and Timestamp Field properties to reference a nested
field with a "/"-delimited path (e.g. "@metadata/index"). A value with no "/" behaves exactly as
before (top-level lookup), so existing flows are unaffected.

When the corresponding Retain property is false, the extracted leaf is removed and any ancestor
object left empty is pruned, so a transient envelope like "@metadata" is not indexed once its
fields are consumed; parents that still have other fields are kept.

Applies across NDJSON, JSON Array, and Single JSON, and all index operations. The NDJSON
raw-bytes fast path is bypassed only when a nested path is configured, since the streaming scan
only matches flat field names.

Builds on NIFI-15985.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`
- Pull request contains [commits signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) with a registered key indicating `Verified` status

### Pull Request Formatting

- Pull Request based on current revision of the `main` branch
- Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `./mvnw clean install -P contrib-check`
  - [x] JDK 21
  - [ ] JDK 25

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java [hunk 1]
```diff
@@ -241,6 +241,7 @@ public class PutElasticsearchJson extends AbstractPutElasticsearch {
             .name("Identifier Field")
             .description("""
                     The name of the field within each document to use as the Elasticsearch document ID. \
+                    A nested field can be referenced with a "/"-delimited path (e.g. "@metadata/id"). \
```
Comment (pvillard31) [inline nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: Could you rebase this branch on the latest main? The validation gate is failing on "branch not up to date", and because of that the full build and test matrix was skipped, so we do not have a green CI signal yet.

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java [hunk 2]
```diff
@@ -1143,30 +1156,168 @@ private String resolveId(final Map<String, Object> contentMap, final String idAt
...
         }
-        final String value = fieldValueToString(contentMap.get(indexField));
+        final String value = fieldValueToString(valueAtPath(contentMap, indexField));
         return StringUtils.isNotBlank(value) ? value : fallbackIndex;
     }
 
+    /**
+     * Whether {@code field} is a {@code /}-delimited nested path rather than a flat field name.
+     */
+    private static boolean isNestedPath(final String field) {
```
Comment (pvillard31) [inline nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java:1230]: A field key that literally contains a slash is now always read as a nested path. Should we note this limitation in the property descriptions so users with such field names are not surprised?

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java [hunk 3]
```diff
@@ -1110,4 +1110,170 @@ void testIndexFieldNullValueFallsBackConsistently() {
...
+        final List<IndexOperationRequest> ops = captureOperations();
+        runner.enqueue("{\"@metadata\":{\"index\":\"docs-single\"},\"msg\":\"x\"}");
+        runner.run();
+
+        assertEquals("docs-single", ops.getFirst().getIndex());
+        assertFalse(ops.getFirst().getFields().containsKey("@metadata"));
+    }
+
+    @Test
+    void testNestedIndexFieldKeepsParentWithRemainingSiblings() {
```
Comment (pvillard31) [inline nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java:1252]: Could we add a test with a 3 level path (for example a/b/c) to lock in the recursive pruning of multiple empty ancestor objects?

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java [hunk 4]
```diff
@@ -1110,4 +1110,170 @@ void testIndexFieldNullValueFallsBackConsistently() {
...
+        final List<IndexOperationRequest> ops = captureOperations();
+        runner.enqueue("{\"@metadata\":{\"index\":\"docs-2026\"},\"msg\":\"hello\"}\n");
+        runner.run();
+
+        assertEquals("docs-2026", ops.getFirst().getIndex());
+        assertTrue(docContent(ops.getFirst()).contains("@metadata"), "nested field retained by default");
+    }
+
+    @Test
+    void testNestedIdentifierFieldNdjsonPrunesEmptyParent() {
```
Comment (pvillard31) [inline nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java:1287]: Should we add a nested identifier test for an Update, Delete, or Upsert operation, so the resolveId nested path is covered as well?

**Review by pvillard31 — CHANGES_REQUESTED** [review]

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java (hunk 2 above)
Comment (agturley) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java:1230]: I've altered it to allow users to escape a backslash to allow a backslash in the path

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java (hunk 3 above)
Comment (agturley) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java:1252]: done

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java (hunk 1 above)
Comment (agturley) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: done

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java (hunk 4 above)
Comment (agturley) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJsonTest.java:1287]: Added testNestedIdentifierFieldUpdateOperationResolvesAndPrunes — it covers the resolveId nested path via an Update op (@metadata/id), verifying the nested ID resolves and the empty parent is pruned. Also added an escaped-slash test on the same path.

**Comment by agturley** [comment]
Do you think the description updates I provided are overkill?

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java [hunk 5]
```diff
@@ -241,6 +241,11 @@ public class PutElasticsearchJson extends AbstractPutElasticsearch {
             .name("Identifier Field")
             .description("""
                     The name of the field within each document to use as the Elasticsearch document ID. \
+                    A nested field can be referenced with a "/"-delimited path: for a document \
```
Comment (pvillard31) [inline nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: These three properties shipped in released 2.10.0 with literal field name semantics, so an existing flow that has a slash in the configured value will now be read as a nested path after an upgrade. I am not an experienced Elasticsearch user and I do not know how common a slash is in these field names, and I agree the likelihood of a problematic value is low, but I think we should still handle this case safely with a property migration that escapes any existing slash or backslash, and add test coverage for that migration. Could we add that here?

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java [hunk 6]
```diff
@@ -267,6 +273,11 @@ public class PutElasticsearchJson extends AbstractPutElasticsearch {
             .name("Index Field")
             .description("""
                     The name of the field within each document to use as the Elasticsearch index name. \
+                    A nested field can be referenced with a "/"-delimited path: for a document \
```
Comment (pvillard31) [inline nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: The escape explanation is repeated in all three property descriptions. Could we move the full explanation into an additionalDetails.md for the processor and have each property description point to the processor documentation for the details instead?

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java [hunk 7]
```diff
@@ -1143,30 +1177,226 @@ private String resolveId(final Map<String, Object> contentMap, final String idAt
...
+        return current;
+    }
+
+    /**
+     * Removes the leaf at the {@code /}-delimited {@code path} within {@code root}, then prunes any
+     * ancestor objects that become empty as a result (the transient envelope often used for routing
+     * metadata, e.g. {@code @metadata}, should not be indexed once its fields are extracted).
+     * Does nothing when the path is absent. A path with no slash is a direct top-level removal.
+     */
+    private static void removeAtPath(final Map<String, Object> root, final String path) {
```
Comment (pvillard31) [inline nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java:1359]: The Map and ObjectNode versions of removeAtPath duplicate the same descend and prune loop. Could we factor the shared walk into one helper so the two cannot drift over time?

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java (hunk 5 above)
Comment (agturley) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: So I started on this and ran into something I want to flag before going further.

The problem is migrateProperties runs on every flow import and restart, and there's no way to tell a 2.10 literal field name like @metadata/id apart from a 2.11 nested path @metadata/id — they're the exact same string. So if I just escape slashes, it also escapes legit nested paths and breaks them. I actually hit this on my test cluster: imported a normal 2.11 flow and every nested-path case quietly stopped working because the migration had rewritten @metadata\/id to @metadata\\/id on import.

The only way I found to make it safe is to rename the three properties (Identifier Field → Identifier Field Path, etc). Then the old name only exists in pre-2.11 flows, so I can escape values found under the old name and leave the new ones alone. That works, but it means renaming three properties everyone's already using.

This is feeling like a lot for the case we're protecting against — a 2.10 flow that happens to have a literal / in the field name, and slashes aren't really common in ES field names to begin with. So I'm inclined to skip the migration and just document it: on upgrade, a field name with a literal / gets read as a nested path now, and you can escape it as \/ if you want the old behavior.

Let me know if you'd rather I do the rename though, happy to if you think the upgrade guarantee is worth it.

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java (hunk 6 above)
Comment (agturley) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: agreed, done

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java (hunk 7 above)
Comment (agturley) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java:1359]: done

File: nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java (hunk 5 above)
Comment (pvillard31) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: @ChrisSamo632 - thoughts on this? You're way more familiar with ES so your feedback would be appreciated if you have the chance

Comment (ChrisSamo632) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: I'm not aware of any restrictions around the use of `/` characters in field names in Elasticsearch or json in general, although I'd imagine the occurrence is relatively infrequent

Trying to think of examples, something like `house_name/number` could be the sort of thing someone might use, but I'd typically opt for an alternative in such a case as I'd see the `/` as unusual (and potentially problematic in some systems/parsers)

For this instance, would a workable option be to add a processor parameter that let's the user configure how a `/` should be treated in field names for this processor? The default could be to have the character treated as a literal, with a new mode that enables the nested object parsing?

Alternatively, would `.` be a better option as it matches the Elasticsearch nested query parsing, but then is further from the nifi record path parsing, which `/` emulates? The same concern would exist, although use of `.` in elasticsearch field names has been discouraged for a long time as it can confuse the Elasticsearch query parser - therefore, the chances of people using `.` litterals on field names seems lower

Comment (ChrisSamo632) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: Thinking about the `.` suggestion a little more - I'd be concerned with someone using that and *not* having the field removed from the json by the processor before it was indexed, which is more likely to run into problems on the Elasticsearch side

So maybe the best idea is to stick with `/` (to be nifi record path-like), and offer a property that uses it as a nested field delimiter (allowing for `\` escaping as needed), or as a literal (the default for backwards compatibility)

What do you think @agturley ?

Comment (agturley) [reply nifi-extension-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchJson.java]: yes i like that idea. i'll adjust and submit.

**Comment by agturley** [comment]
Sorry about the delay in responding to this. I was out last week and was not able to do any updates.
Anyways, There's now a "Field Path Mode" property with two options:

**Literal Field Name (default):**  the Identifier/Index/Timestamp Field value is the exact name of a top-level field, exactly like before. / and \ are just regular characters.
**Nested Field Path:** the value is a /-delimited path into nested objects (e.g. @metadata/id), with \/ to escape a literal slash.
Because it defaults to Literal Field Name, existing flows keep working with no change and no migration is needed — nesting is purely opt-in, so there's no risk of an existing field name with a / being reinterpreted on upgrade.

I tested it end to end against a cluster (nested id/index/timestamp extraction, deep paths, escaped slashes, and plain pass-through) and it all behaves as expected. I also added an example for each mode to the processor's Additional Details, and tests covering both the literal default and the nested behavior.

**Review by pvillard31 — APPROVED** [review]
Thanks for making the changes, the latest LGTM. @ChrisSamo632 - anything on your side?

**Comment by ChrisSamo632** [comment]
LGTM 👍

Thanks for the improvements @agturley , and the detailed review @pvillard31

**Review by ChrisSamo632 — APPROVED** [review]
